### PR TITLE
rubocop

### PR DESCRIPTION
**Pull request by grosser**
@bquorning

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,197 @@
+AllCops:
+  DisplayCopNames: true
+  TargetRubyVersion: 2.3
```
Comment (bquorning) [inline .rubocop.yml]: There should be a matching ruby version restriction in the gemspec. Since this is a public gem, I’d say 2.2 or even as low as 2.1. Also .travis.yml should probably test every minor version we claim to be compatible with.

File: .rubocop.yml [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+
+Style/ClassAndModuleChildren:
+  Enabled: false
+
+# %w[] shows that it will return an array
+Style/PercentLiteralDelimiters:
+  PreferredDelimiters:
+    '%i': '[]'
+    '%w': '[]'
+    '%W': '[]'
```
Comment (bquorning) [inline .rubocop.yml]: These are the default values in newer versions of Rubocop.

File: gemfiles/common.rb [hunk 3]
```diff
@@ -1,6 +1,8 @@
+# frozen_string_literal: true
 source 'https://rubygems.org'
 
 gem 'activerecord-jdbcmysql-adapter', '~> 1.2', platforms: :jruby
 gem 'byebug', platforms: :ruby
+gem 'rubocop', platforms: :ruby
```
Comment (bquorning) [inline gemfiles/common.rb]: Rubocop has a history of breaking changes between minor releases. I think we should lock to `~> 0.49.1` for now.

File: Rakefile [hunk 4]
```diff
@@ -10,3 +11,8 @@ Rake::TestTask.new(:test) do |test|
 end
 
 task default: 'wwtd:local'
+
+desc "Run rubocop"
+task :rubocop do
+  sh "rubocop"
+end
```
Comment (bquorning) [inline Rakefile]: Instead of writing your own task, use the built-in one:
```rb
require 'rubocop/rake_task'

RuboCop::RakeTask.new
```

File: kasket.gemspec [hunk 5]
```diff
@@ -21,5 +22,5 @@ Gem::Specification.new do |s|
   s.add_development_dependency("minitest-rg")
   s.add_development_dependency("timecop")
 
-  s.files        = Dir.glob("lib/**/*") + %w(README.md)
+  s.files = Dir.glob("lib/**/*") + %w[README.md]
```
Comment (bquorning) [inline kasket.gemspec]: Perhaps move this line to right under `s.license` and keep the spaces.

File: lib/kasket/configuration_mixin.rb [hunk 6]
```diff
@@ -61,16 +63,18 @@ def kasket_indices
       result.uniq
     end
 
-    def has_kasket_index_on?(sorted_attributes)
+    def has_kasket_index_on?(sorted_attributes) # rubocop:disable Style/PredicateName
```
Comment (bquorning) [inline lib/kasket/configuration_mixin.rb]: Can we just disable  Style/PredicateName globally?

File: lib/kasket/configuration_mixin.rb [hunk 7]
```diff
@@ -61,16 +63,18 @@ def kasket_indices
...
       kasket_indices.include?(sorted_attributes)
     end
 
-    def has_kasket(options = {})
```
Comment (bquorning) [inline lib/kasket/configuration_mixin.rb]: You’re changing the method signature on a public method here, right? Maybe `def has_kasket(_options = {})` instead?

File: lib/kasket/visitor.rb [hunk 8]
```diff
@@ -95,10 +96,10 @@ def visit_Arel_Nodes_In(node, *_)
 
     def visit_Arel_Nodes_Equality(node, *_)
       right = case node.right
-      when false then 0 # This should probably be removed when Rails 3.2 is not supported anymore
-      when nil   then nil
-      else
-        visit(node.right)
+              when false then 0 # This should probably be removed when Rails 3.2 is not supported anymore
```
Comment (bquorning) [inline lib/kasket/visitor.rb]: 2 space indentation should suffice here.

File: lib/kasket/write_mixin.rb [hunk 9]
```diff
@@ -136,7 +138,7 @@ def self.included(model_class)
       model_class.after_destroy :kasket_after_destroy
       model_class.after_commit :kasket_after_commit_dummy
 
-      if ActiveRecord::VERSION::MAJOR == 3 || (ActiveRecord::VERSION::MAJOR == 4 && ActiveRecord::VERSION::MINOR == 0)
+      if ActiveRecord::VERSION::MAJOR == 3 || (ActiveRecord::VERSION::MAJOR == 4 && ActiveRecord::VERSION::MINOR.zero?)
```
Comment (bquorning) [inline lib/kasket/write_mixin.rb]: Please disable whatever cop suggested to use `.zero?`.

File: test/dirty_test.rb [hunk 10]
```diff
@@ -6,7 +7,7 @@
   def assert_cleared
     post = Post.first
 
-    Post.cache { pots = Post.find(post.id) }
+    Post.cache { post = Post.find(post.id) }
```
Comment (bquorning) [inline test/dirty_test.rb]: ooh

File: test/parser_test.rb [hunk 11]
```diff
@@ -1,14 +1,15 @@
...
 describe Kasket::QueryParser do
   def parse(options)
     scope = Post
-    options.each do |k,v|
+    options.each do |k, v|
       scope = case k
-      when :conditions then scope.where(v)
-      else
-        scope.send(k, v)
+              when :conditions then scope.where(v)
```
Comment (bquorning) [inline test/parser_test.rb]: Again, too much indentation for my taste. I’m sure there’s a configuration options for allowing 2-space indentation here.

File: .rubocop.yml [hunk 12]
```diff
@@ -0,0 +1,197 @@
+AllCops:
+  DisplayCopNames: true
+  TargetRubyVersion: 2.3
+  Exclude:
+    - db/schema.rb
+    - vendor/**/*
+    - gemfiles/vendor/**/*
+    - tmp/**/*
+
+LineLength:
```
Comment (bquorning) [inline .rubocop.yml]: In my .rubocop.yml files, all cops are namespaced and listed alphabetically. Can we do the same here?

File: .rubocop.yml [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  Enabled: false
+
+Layout/CommentIndentation:
+  Enabled: false
+
+Style/StringLiterals:
+  Enabled: false
+
+HashSyntax:
+   EnforcedStyle: ruby19
```
Comment (bquorning) [inline .rubocop.yml]: I think this is also the default value.

File: .rubocop.yml [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+Style/SignalException:
+  Enabled: false
+
+Layout/SpaceInsideHashLiteralBraces:
+  Enabled: false
```
Comment (bquorning) [inline .rubocop.yml]: Please enable this cop.

File: .rubocop.yml [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+Style/DoubleNegation:
+  Enabled: false
+
+Style/Documentation:
+  Enabled: false
+
+Style/Next:
+  Enabled: false
+
+Metrics/CyclomaticComplexity:
```
Comment (bquorning) [inline .rubocop.yml]: You have this cop disabled twice.

File: .rubocop.yml [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+
+Style/VariableNumber:
+  Enabled: false
+
+Style/TernaryParentheses:
+  Enabled: false
+
+# should use, but is currently showing false positives / bugs
+# https://github.com/bbatsov/rubocop/issues/3551
+# https://github.com/bbatsov/rubocop/issues/3561
```
Comment (bquorning) [inline .rubocop.yml]: These bugs were fixed almost a year ago.

File: .rubocop.yml (hunk 1 above)
Comment (grosser) [inline .rubocop.yml]: then I'd get different errors reported ... trying to run on a single version and use that as long as other versions are passing

File: .rubocop.yml (hunk 14 above)
Comment (grosser) [inline .rubocop.yml]: was even more diff, can come later

File: lib/kasket/configuration_mixin.rb (hunk 7 above)
Comment (grosser) [inline lib/kasket/configuration_mixin.rb]: wanted to change it anyway since it's bad to have anyone call it with options and then have them ignored

**Comment by grosser** [comment]
all updated

File: lib/kasket/configuration_mixin.rb (hunk 7 above)
Comment (bquorning) [inline lib/kasket/configuration_mixin.rb]: You really don't want breaking changes in this PR.

File: .rubocop.yml (hunk 1 above)
Comment (bquorning) [inline .rubocop.yml]: I don't remember the details, but using 2.3 might cause Rubocop to enforce some rules that are not compatible with 2.2.

Comment (grosser) [inline .rubocop.yml]: we got tests for that :)

File: lib/kasket/configuration_mixin.rb (hunk 7 above)
Comment (grosser) [inline lib/kasket/configuration_mixin.rb]: ... ok added it back ...

File: .rubocop.yml (hunk 1 above)
Comment (bquorning) [inline .rubocop.yml]: True. Should we add ruby version to the gemspec?

File: .rubocop.yml [hunk 17]
```diff
@@ -0,0 +1,154 @@
+AllCops:
+  DisplayCopNames: true
+  TargetRubyVersion: 2.3
+  Exclude:
+    - db/schema.rb
```
Comment (bquorning) [inline .rubocop.yml]: No such file.

File: .rubocop.yml [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+  Enabled: false
+
+Layout/IndentHash:
+  Enabled: false
+
+Layout/MultilineMethodCallIndentation:
+  Enabled: false
+
+Layout/MultilineOperationIndentation:
+  Enabled: false
```
Comment (bquorning) [inline .rubocop.yml]: Can we please change part of the `Layout` configuration (current lines 25–35) into these?
```yml
Layout/FirstParameterIndentation:
  EnforcedStyle: consistent

Layout/IndentArray:
  EnforcedStyle: consistent

Layout/IndentHash:
  EnforcedStyle: consistent

Layout/MultilineMethodCallIndentation:
  EnforcedStyle: indented

Layout/MultilineOperationIndentation:
  EnforcedStyle: indented
```

File: lib/kasket.rb [hunk 19]
```diff
@@ -14,7 +15,7 @@ module Kasket
   autoload :SelectManagerMixin,     'kasket/select_manager_mixin'
   autoload :RelationMixin,          'kasket/relation_mixin'
 
-  CONFIGURATION = {:max_collection_size => 100, :write_through => false}
+  CONFIGURATION = { max_collection_size: 100, write_through: false } # rubocop:disable Style/MutableConstant
```
Comment (bquorning) [inline lib/kasket.rb]: I’d prefer if we would avoid inline `rubocop:disable`s.

File: .rubocop.yml (hunk 1 above)
Comment (grosser) [inline .rubocop.yml]: Yes, should be >= least tested version ... but that can be another PR

File: lib/kasket.rb (hunk 19 above)
Comment (grosser) [inline lib/kasket.rb]: in test we modify the constant ... wanted a quick PR to fix all the simple things first

**Comment by grosser** [comment]
updated

**Comment by bquorning** [comment]
:+1:
